apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+    for (int i = 0; i < (int)order.size() - (int)outputs.size(); ++i) {
+        bool is_output = false;
+        for (const Function &f : outputs) {
+            if (order[i] == f.name()) {
+                is_output = true;
+                break;
+            }
+        }
+        if (is_output) {
+            // Should not inline output Func
+            debug(5) << "Skip inlining " << order[i] << " since it is an output\n";
+            continue;
+        }
+        std::string caller = is_func_called_element_wise(order, i, env);
+        if (!caller.empty()) {
+            inlined = true;
+            debug(4) << "Inline function \"" << order[i] << "\" since it is called only by "
+                     << caller << " in element-wise manner\n";
+            internal_assert(order[i] != caller);
+            Function f1 = get_element(env, order[i]);
+            f1.schedule().store_level().lock();
+            inline_function(env.at(caller), f1);
+        }
+    }
+    return inlined;
+}
+
+std::vector<int> get_int_bounds(const Box &bounds) {
+    std::vector<int> int_bounds;
+    int_bounds.reserve(bounds.size());
+    for (int i = 0; i < (int)bounds.size(); i++) {
+        Interval interval = bounds[i];
+        Expr extent = simplify(interval.max - interval.min + 1);
+        extent = simplify(substitute_var_estimates(extent));
+        const int64_t *extent_int = as_const_int(extent);
+        user_assert(extent_int != nullptr) <<
+            "extent:" << extent << " is not constant.\n";
+        int_bounds.push_back(*extent_int);
+    }
+    return int_bounds;
+}
+
+std::vector<int> get_rvar_bounds(const std::vector<ReductionVariable> &rvars) {
+    std::vector<int> rvar_bounds;
+    rvar_bounds.reserve(rvars.size());
+    for (int arg_id = 0; arg_id < (int)rvars.size(); arg_id++) {
+        Expr extent = simplify(substitute_var_estimates(rvars[arg_id].extent));
+        const int64_t *extent_int = as_const_int(extent);
+        user_assert(extent_int != nullptr) <<
+            "extent:" << extent << " is not constant.\n";
+        rvar_bounds.push_back(*extent_int);
+    }
+    return rvar_bounds;
+}
+
+void reorder_storage(Func func,
+                     const std::vector<Var> &all_vars,
+                     std::ostringstream &schedule_source) {
+    func.reorder_storage(all_vars);
+    schedule_source << "    .reorder_storage(";
+    for (int i = 0; i < (int)all_vars.size(); i++) {
+        schedule_source << all_vars[i].name();
+        if (i != (int)all_vars.size() - 1) {
+            schedule_source << ",";
+        }
+    }
+    schedule_source << ")\n";
+}
+
+void reorder_storage(Stage stage,
+                     const std::vector<Var> &all_vars,
+                     std::ostringstream &schedule_source) {
+    // Can't reorder storage of a stage.
+}
+
+int natural_vector_size(const Target &target, const Type &t) {
+    const bool is_integer = t.is_int() || t.is_uint();
```

Comment:
`Target::natural_vector_size` already exists -- why aren't we using it? (If it doesn't handle AVX512 we should augment it to do so rather than reinvent the wheel here)

Reply:
Now we use target.natural_vector_size (it does handle AVX512). The issue was that in my test the target is unknown.